official/resnet/imagenet_main.py

```diff
@@ -144,6 +143,12 @@ def input_fn(is_training, data_dir, batch_size, num_epochs=1,
   # Convert to individual records
   dataset = dataset.flat_map(tf.data.TFRecordDataset)
 
+  # Currently, if we are using multiple GPUs, we can't pass in uneven batches.
+  # This ensures that we cut off the uneven images at the end of each epoch
+  # (there are fifteen leftovers), rather than trying and failing to split
```

Comment:
What does "there are fifteen leftovers" mean? Perhaps say ("e.g. there are 15 leftovers when the batch size is 32").

That said though, some thoughts:

- Why doesn't `cifar10_main.py` need the same truncation logic?

- This silent truncation of the dataset is subtle and may be worth bubbling up to the caller (e.g., by adding an explicit `truncate` argument to `input_fn`).

- (Non-actionable note): `resnet.process_record_dataset` seems to `batch` after `repeat(num_epochs)`, which means that one could have avoided the truncation (or could have truncated less) depending on what `num_epochs` is. For example, for `num_epochs = 15`, no truncation is needed :). I point this out not because I think those 15 examples are particularly important, but just that these interactions are quite subtle and add another source of difference. For example, the comparsion between a batch size of 128 and a batch size of 1024 may also be influenced by the difference in the ~130 examples seen by one and not the other?

Reply:
Yeah, this is not my favorite line. Right now, what happens is there are 1281167 training images; if you run with one epoch, you get 1281167/32 batches, which comes to ~40K batches with 15 leftover. Running on one GPU, that doesn't matter. But with the current replicate_model_fn, everything breaks at the end of the epoch when the split process can't divide 15 examples evenly over 4 or 8 GPUs. This is the current solution @isaprykin suggested, pending a decision for replicate_model_fn more generally.

To your specific points--
1. On the one hand, yes, it should be in cifar10 if we want to run that multi-GPU in the same way. However, because cifar happens to have a nicely divisible number of images, it's not practically a likely issue. I therefore wanted to err on the side of not including a hacky line for a case that wasn't likely to arise, but maybe it's better to just have in both.

2. Or maybe just do it for multi-GPU for now? It seems like unnecessary complication to do that for the general case when it is a half-bandaid for now. The reason I didn't do that is that right now that function is not multi-GPU aware, and I hate to introduce a third place where we have to do special things for multi-GPU.

3. Throwing that one back to @isaprykin , who is deciding what the solution should be more generally :)

Thoughts?

Reply:
2. sounds good.

Thanks for the details. (Perhaps worth capturing this in a comment? Your call)